lib/puppet/provider/package/pip.rb

```diff
@@ -166,38 +200,68 @@ def latest_with_old_pip
       execpipe command_and_options do |process|
         process.collect do |line|
           # PIP OUTPUT: Using version 0.10.1 (newest of versions: 1.2.3, 4.5.6)
-          if line =~ /Using version (.+?) \(newest of versions/
-            return $1
+          if line =~ /Using version .+? \(newest of versions: (.+?)\)/
+            versionList = $1.split(', ').sort do |x,y|
+              Puppet::Util::Package.versioncmp(x, y)
+            end
+            return versionList
           end
         end
-        return nil
       end
+      return []
+    end
+  end
+
+  # Finds the most suitable version available in a given range
+  def best_version(should_range)
+    included_available_versions = []
+    available_versions.each do |version|
+      version = PIP_VERSION.parse(version)
+      included_available_versions.push(version) if should_range.include?(version)
     end
+
+    included_available_versions.sort!
+    return included_available_versions.last unless included_available_versions.empty?
+
+    Puppet.debug("No available version for package #{@resource[:name]} is included in range #{should_range}")
+    should_range
   end
 
   # Install a package.  The ensure parameter may specify installed,
   # latest, a version number, or, in conjunction with the source
   # parameter, an SCM revision.  In that case, the source parameter
   # gives the fully-qualified URL to the repository.
-
   def install
     command = resource_or_provider_command
     self.class.validate_command(command)
 
+    should = @resource[:ensure]
     command_options = %w{install -q}
     command_options +=  install_options if @resource[:install_options]
     if @resource[:source]
-      if String === @resource[:ensure]
-        command_options << "#{@resource[:source]}@#{@resource[:ensure]}#egg=#{@resource[:name]}"
+      if String === should
+        command_options << "#{@resource[:source]}@#{should}#egg=#{@resource[:name]}"
       else
         command_options << "#{@resource[:source]}#egg=#{@resource[:name]}"
       end
     else
-      case @resource[:ensure]
-      when String
-        command_options << "#{@resource[:name]}==#{@resource[:ensure]}"
+      case should
       when :latest
         command_options << "--upgrade" << @resource[:name]
+      when String
+        begin
+          should_range = PIP_VERSION_RANGE.parse(should, PIP_VERSION)
+          should = best_version(should_range)
+
+          if should.is_a?(PIP_VERSION_RANGE)
```

Comment:
hmmm, this feels a bit weird, we are trying to find the best versions, and if we cannot do that we return a range, and then we pass that range to pip command?

But i don't think we get PIP_VERSION_RANGE


```
❯ bx puppet resource package 'colorize' ensure='>3' provider='pip3'


=> #<Puppet::Util::Package::Version::Range::Gt:0x00007f88c0523bb8 @version=3>
[6] pry(#<Puppet::Type::Package::ProviderPip3>)> should.class
=> Puppet::Util::Package::Version::Range::Gt
```

Reply:
After discussing the issue, we agreed on previous implementation with a comment explaining the reason. Also the MinMax scenario was discovered and added.